etcdserver/raft.go

```diff
@@ -203,6 +212,67 @@ func (r *raftNode) leadElectedTime() time.Time {
 	return r.lt
 }
 
+func (r *raftNode) retryMsgProp(m raftpb.Message) {
+	if m.Type != raftpb.MsgProp {
+		return
+	}
+	r.mu.Lock()
+	defer r.mu.Unlock()
+	r.retryMsgProps = append(r.retryMsgProps, m)
+}
+
+// the loop holds the proposals waiting for retry, and proposes them
+// when there is chance for them to be committed. If it finds out local raft
+// is isolated from the majority for a long time, it drops buffered proposals
+// because it cannot commit them anyway.
+func (r *raftNode) retryMsgPropLoop(t <-chan time.Time) {
+	// the number of ticks that fails to retry proposals
+	var failedTicks int
+	for {
+		select {
+		case <-t:
+			lead := atomic.LoadUint64(&r.lead)
+			// retry proposal cannot succeed when there is no leader
+			if lead == raft.None {
+				failedTicks++
+				break
+			}
+			failedTicks = 0
+
+			r.mu.Lock()
+			msgs := r.retryMsgProps
+			r.retryMsgProps = nil
+			r.mu.Unlock()
+			ctx, cancel := context.WithTimeout(context.Background(), time.Second)
```

Comment:
> Moreover, the default election timeout is 1s, so even if it is context deadline exceeded here, it could catch up the next ticker.

I do not understand this. It seems that we cannot assume anything around the default case.